server.js

```diff
@@ -48,56 +51,98 @@ const CONFIG = {
       sub3: `${SPECIAL_DELIM}header${SPECIAL_SEP}x-source${SPECIAL_DELIM}`,
       cu: `${SPECIAL_DELIM}header${SPECIAL_SEP}x-target-url${SPECIAL_DELIM}`
     }
-  }
+  },
+  weather_conditions: {
+    url: process.env["WEATHER_CONDITIONS_URL"],
+    query: {
+      apikey: process.env["WEATHER_KEY"],
+    }
+  },
+  weather_location: {
+    url: process.env["WEATHER_LOCATION_URL"],
+    query: {
+      // Consumers should pass a `q` parameter in the URL.
+      apikey: process.env["WEATHER_KEY"],
+    }
+  },
 };
 
+/**
+ * Constructs the URL to be fetched via this proxy. First, we merge pre-defined
+ * query parameters from `options` (if any) with any query paramters passed in
+ * via the URL. Then we make partner-specific adjustments to the URL. Finally,
+ * we perform any required special handling on query parameters, such as
+ * encoding URIs.
+ *
+ * @param {Request} req
+ *   The request from the client.
+ * @param {object} options
+ *   The object in CONFIG above corresponding to the endpoint requested by the
+ *   client. For example, if the client called
+ *   https://<SERVER_URL>/cid/amzn_2020_a1, `options` should be the contents of
+ *   CONFIG.amzn_2020_a1.
+ */
 const createTarget = (req, options) => {
   let query = [];
-  if (options.query) {
-    // Clone the object for single use inside this method.
-    options = Object.assign({}, options);
-    options.query = Object.assign({}, options.query);
-
-    let XTargetURL = req.headers["x-target-url"];
-    // Support the eBay campaign.
-    if (XTargetURL && XTargetURL.startsWith("https://www.ebay.")) {
-      options.query.sub1 = "ebay";
+
+  // Clone the object for single use inside this method.
+  options = Object.assign({}, options);
+  options.query = Object.assign({}, options.query);
+  if (req.query) {
+    // Merge the query parameters passed to the proxy with those pre-defined in
+    // options.query.
+    Object.assign(options.query, req.query);
+  }
+
+  let XTargetURL = req.headers["x-target-url"];
+  // Support the eBay campaign.
+  if (XTargetURL && XTargetURL.startsWith("https://www.ebay.")) {
+    options.query.sub1 = "ebay";
+  }
+
+  if (options.url == process.env["WEATHER_CONDITIONS_URL"]) {
+    // The weather conditions API requires a dynamic key to be in the URL path
+    // rather than in the query paramters.
+    if (options.query.locationKey) {
+      options.url = `${options.url}${options.query.locationKey}.json`;
+    } else {
+      return;
```

Comment:
Why do this?

Reply:
If the client is trying to access the weather conditions API but didn't provide a location key, we can't fetch any useful data. There's not much we can do but return.

Reply:
Alright, makes sense. Can you throw an error instead? We don't want the request to continue either in that case, right? Throwing an error will make the request fail with status code 500.